apiserver/uniter/subordinaterelationwatcher.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package uniter
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/utils/set"
+	"gopkg.in/juju/charm.v6-unstable"
+
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/watcher"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+type subRelationsWatcher struct {
+	catacomb      catacomb.Catacomb
+	backend       *state.State
+	app           *state.Application
+	principalName string
+
+	// Maps relation keys to whether that relation should be
+	// included. Needed particularly for when the relation goes away.
+	relations map[string]bool
+	out       chan []string
+}
+
+// newSubordinateRelationsWatcher creates a watcher that will notify
+// about relation lifecycle events for subordinateApp, but filtered to
+// be relevant to a unit deployed to a container with the
+// principalName app. Global relations will be included, but only
+// container-scoped relations for the principal application will be
+// emitted - other container-scoped relations will be filtered out.
+func newSubordinateRelationsWatcher(backend *state.State, subordinateApp *state.Application, principalName string) (
+	state.StringsWatcher, error,
+) {
+	w := &subRelationsWatcher{
+		backend:       backend,
+		app:           subordinateApp,
+		principalName: principalName,
+		relations:     make(map[string]bool),
+		out:           make(chan []string),
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+	return w, errors.Trace(err)
+}
+
+func (w *subRelationsWatcher) loop() error {
+	defer close(w.out)
+	relationsw := w.app.WatchRelations()
+	if err := w.catacomb.Add(relationsw); err != nil {
+		return errors.Trace(err)
+	}
+	var (
+		sentInitial bool
+		outputData  []string
+		out         chan []string
+
+		currentRelations = set.NewStrings()
+	)
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case out <- outputData:
+			sentInitial = true
+			currentRelations = set.NewStrings()
+			outputData = nil
+			out = nil
+		case newRelations, ok := <-relationsw.Changes():
+			if !ok {
+				return watcher.EnsureErr(relationsw)
+			}
+			for _, relation := range newRelations {
+				if currentRelations.Contains(relation) {
+					continue
+				}
+				shouldSend, err := w.shouldSend(relation)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if shouldSend {
+					currentRelations.Add(relation)
```

Comment:
My concern with doing that was that in a pathological case when there's a problem with the downstream so the outgoing messages aren't getting picked up and the current relations set is getting bigger and bigger, then each time you get one more relation you loop across all of the accumulated set to recreate the output slice, so it becomes O(n^2).